src/ucp/rma/amo_offload.c

```diff
@@ -0,0 +1,155 @@
+    uct_atomic_op_t op                   = req->send.amo.uct_op;
+    uint64_t *result64                   = req->send.buffer;
+    uint32_t *result32                   = req->send.buffer;
+    ucs_status_t status;
+    uct_rkey_t tl_rkey;
+
+    req->send.lane = spriv->super.lane;
+    tl_rkey        = ucp_rma_request_get_tl_rkey(req, spriv->super.rkey_index);
+
+    if ((op_id != UCP_OP_ID_AMO_POST) &&
+        !(req->flags & UCP_REQUEST_FLAG_PROTO_INITIALIZED)) {
+        req->send.state.uct_comp.count  = 1;
+        req->send.state.uct_comp.status = UCS_OK;
+        req->send.state.uct_comp.func   = ucp_proto_amo_completed;
+        req->flags                     |= UCP_REQUEST_FLAG_PROTO_INITIALIZED;
+    }
+
+    if (op_size == sizeof(uint64_t)) {
+        if (op_id == UCP_OP_ID_AMO_POST) {
+            status = UCS_PROFILE_CALL(uct_ep_atomic64_post,
+                                      ep->uct_eps[req->send.lane], op, value,
+                                      remote_addr, tl_rkey);
+        } else if (op_id == UCP_OP_ID_AMO_FETCH) {
+            status = UCS_PROFILE_CALL(uct_ep_atomic64_fetch,
+                                      ep->uct_eps[req->send.lane], op, value,
+                                      result64, remote_addr, tl_rkey,
+                                      &req->send.state.uct_comp);
+        } else {
+            ucs_assert(op_id == UCP_OP_ID_AMO_CSWAP);
+            status = UCS_PROFILE_CALL(uct_ep_atomic_cswap64,
+                                      ep->uct_eps[req->send.lane], value,
+                                      *result64, remote_addr, tl_rkey, result64,
+                                      &req->send.state.uct_comp);
+        }
+    } else {
+        ucs_assert(op_size == sizeof(uint32_t));
+        if (op_id == UCP_OP_ID_AMO_POST) {
+            status = UCS_PROFILE_CALL(uct_ep_atomic32_post,
+                                      ep->uct_eps[req->send.lane], op, value,
+                                      remote_addr, tl_rkey);
+        } else if (op_id == UCP_OP_ID_AMO_FETCH) {
+            status = UCS_PROFILE_CALL(uct_ep_atomic32_fetch,
+                                      ep->uct_eps[req->send.lane], op, value,
+                                      result32, remote_addr, tl_rkey,
+                                      &req->send.state.uct_comp);
+        } else {
+            ucs_assert(op_id == UCP_OP_ID_AMO_CSWAP);
+            status = UCS_PROFILE_CALL(uct_ep_atomic_cswap32,
+                                      ep->uct_eps[req->send.lane], value,
+                                      *result32, remote_addr, tl_rkey, result32,
+                                      &req->send.state.uct_comp);
+        }
+    }
+
+    if (status == UCS_INPROGRESS) {
+        return UCS_OK;
+    }
+
+    /* Complete for UCS_OK and unexpected errors */
+    if (status != UCS_ERR_NO_RESOURCE) {
+        ucp_request_complete_send(req, status);
+    }
+
```

Comment:
- check if can use common protocol function to handle status
- if there is error which is not NO_RESOURCE: need to call proto_request_abort() and return UCS_OK